src/main/java/com/commercetools/sync/commons/utils/ClientConfigurationUtils.java

```diff
@@ -45,76 +56,74 @@
 
-    /**
-     * Creates a {@link BlockingSphereClient} with a default {@code timeout} value of 30 seconds.
-     *
-     * @param clientConfig the client configuration for the client.
-     * @return the instantiated {@link BlockingSphereClient}.
-     */
-    public static SphereClient createClient(@Nonnull final SphereClientConfig clientConfig) {
-        return createClient(clientConfig, DEFAULT_TIMEOUT, DEFAULT_TIMEOUT_TIME_UNIT);
+    protected static SphereClient createSphereClient(@Nonnull final SphereClientConfig clientConfig) {
+        final HttpClient httpClient = getHttpClient();
+        final SphereAccessTokenSupplier tokenSupplier =
+            SphereAccessTokenSupplier.ofAutoRefresh(clientConfig, httpClient, false);
+        return SphereClient.of(clientConfig, httpClient, tokenSupplier);
+    }
+
+    protected static SphereClient decorateSphereClient(
+        @Nonnull final SphereClient underlyingClient,
+        final long maxRetryAttempt,
+        @Nonnull final Function<RetryContext, Duration> durationFunction) {
+
+        final SphereClient retryClient = withRetry(underlyingClient, maxRetryAttempt, durationFunction);
+        return withLimitedParallelRequests(retryClient);
     }
 
     /**
      * Gets an asynchronous {@link HttpClient} to be used by the {@link BlockingSphereClient}.
-     * Client is created during first invocation and then cached.
      *
      * @return {@link HttpClient}
      */
     private static synchronized HttpClient getHttpClient() {
-        if (httpClient == null) {
-            final AsyncHttpClient asyncHttpClient =
-                new DefaultAsyncHttpClient(
-                    new DefaultAsyncHttpClientConfig.Builder()
-                                                    .setHandshakeTimeout((int) DEFAULT_TIMEOUT)
-                                                    .build());
-            httpClient = AsyncHttpClientAdapter.of(asyncHttpClient);
-        }
-        return httpClient;
+        final AsyncHttpClient asyncHttpClient =
+            new DefaultAsyncHttpClient(
+                new DefaultAsyncHttpClientConfig.Builder()
+                                                .setHandshakeTimeout((int) DEFAULT_TIMEOUT)
+                                                .build());
+        return AsyncHttpClientAdapter.of(asyncHttpClient);
     }
 
-    private static SphereClient withRetry(final SphereClient delegate) {
-        final int maxAttempts = 5;
-        final RetryAction scheduledRetry = RetryAction
-            .ofScheduledRetry(maxAttempts, context -> calculateVariableDelay(context.getAttempt()));
+    private static SphereClient withRetry(
+        @Nonnull final SphereClient delegate,
+        long maxRetryAttempt,
+        @Nonnull final Function<RetryContext, Duration> durationFunction) {
+
+        final RetryAction scheduledRetry = RetryAction.ofScheduledRetry(maxRetryAttempt, durationFunction);
         final RetryPredicate http5xxMatcher = RetryPredicate
```

Comment:
Should it retry for all 5xx errors? Like we have here: https://github.com/sphereio/sphere-node-sdk/blob/master/src/coffee/repeater-task-queue.coffee#L78. Additionally, we also have retry on some particular error statuses, maybe you can check them as well if they fit here.

Reply:
Actually, in commercetools docs, retry is suggested only for 503 and 409 (after resolving the version), so I am not sure when to retry. It is also interesting that the new node SDK is retrying only for 503. In this util class, in addition to 503, we are retrying for 502 and 504. Let me know if you know more details about this topic.

Reply:
I found an issue for the topic: https://github.com/commercetools/commercetools-sync-java/issues/208, I need to check the findings that we have before.

Reply:
I have checked the HTTP client library-`asynchttpclient`, and those values are is the defaults, comes from the file called, `ahc-defaults.properties`, I guess using 30 seconds handshake is not a big improvement here, not sure why it's needed, also in 208 its suggested less timeouts than defaults, which does not look right to me.

```` java
        final DefaultAsyncHttpClientConfig build = new DefaultAsyncHttpClientConfig.Builder()
            .build();
        System.out.println(build.getReadTimeout()); // 60000
        System.out.println(build.getConnectTimeout()); // 5000
        System.out.println(build.getHandshakeTimeout()); // 10000
        System.out.println(build.getRequestTimeout());; // 60000
````

Reply:
> I have checked the HTTP client library-`asynchttpclient`, and those values are is the defaults, comes from the file called, `ahc-defaults.properties`, I guess using 30 seconds handshake is not a big improvement here, not sure why it's needed, also in 208 its suggested less timeouts than defaults, which does not look right to me.
> 
> ```java
>         final DefaultAsyncHttpClientConfig build = new DefaultAsyncHttpClientConfig.Builder()
>             .build();
>         System.out.println(build.getReadTimeout()); // 60000
>         System.out.println(build.getConnectTimeout()); // 5000
>         System.out.println(build.getHandshakeTimeout()); // 10000
>         System.out.println(build.getRequestTimeout());; // 60000
> ```

Are you talking about the retry for 5xx? Anyway, seems like you're right with the documentation saying you should retry only on 503 😕 So you can choose - either follow the official docs and retry only on 503 or retry on all 5xx errors 🙂  I don't know much about this topic, I guess you can create a support ticket to ask for sure.

Reply:
I would go with a retry on any 5xx errors as it might be just a temporal issue on one of the API nodes only. Defaults for read/connect/handshake timeouts seem fine.

Reply:
I am not sure about the topic, and not sure why docs are suggesting only for 503. 

What about the 500? We don't know what is the problem that returns 500? If it's a node related problem so not a problem, it could be also the opposite, retrying all 5xx could also harm the platform if it's a bug, right ?

Reply:
Maybe this case should be retried but not more than once, if it's a bug it will not be different, if not it might be resolved. WDYT?